Add support for EgressIP and ExternalTrafficPolicy=Local to coexist

tssurya opened the pull request:
**- What this PR does and why is it needed**

This PR fixes the broken interaction between egressIPs and ETP=local features.
Currently if an EIP served pod is on a different node than the egressNode and
it is also a backend to the ETP=local service type, then the reply packet destined
for the ingress service traffic served by the EIP pod get's re-routed to the egressNode
using the reRoute policies we add for EIPs. This breaks the service traffic.

This PR implements the solution we did via https://issues.redhat.com/browse/FDP-42
to ensure these features work together seamlessly.

Solution:

Step1: Add a qos-rule to the node switches:

`from-lport   103 (ip4.src == $a4548040316634674295 && ct.trk && ct.rpl) mark=42`
where `a4548040316634674295` is the address set that contains all EIP pods.

Step2: Add new LRP at 102 to match on mark and simply "allow" thus skipping the re-routes
LRP:

sample:
```
102 pkt.mark == 42 allow
```

**However we must do the mark changes ONLY for local zone pods not for remote zone pods, else it breaks traffic since mark doesn't persist outside the local node.**

**- Special notes for reviewers**

- The logic to add QoS Rules has been added to the syncNamespaces/initClusterPolicies function so that on start all nodes get this rule
- It has also being added to the addNodeEvent handler so that when new node is added we get this rule
- It has not been added to updateNode event since there is nothing to do during node updates
- It has also not been added to the deleteNode event handler since switch goes away and rules will also go away unless it is still being referenced by another switch
- EgressSVC feature has not been added to this mix because there the reply traffic should be going out with SVC VIP so unsure what the expected behaviour for that feature is; but in the future if that is something that needs to be fixed it should be easy to do that the same way
- since its new LRP/QoS add it will take effect on upgrades

**- How to verify it**

1. All changes to existing unit tests already cover the unit testing around this
2. E2E has been added for cross functionality testing between services and egress ips.

The `2024-04-25T14:09:40.7078121Z [0mLoad Balancer Service Tests with MetalLB [0m[1mShould ensure load balancer service works when ETP=local and backend pods are also egressIP served pods[0m` is running on both v4 and v6 lanes for LGW gateway mode and here the egressNode is different from the node where the pod lives

AND `should work on secondary node interfaces for ETP=local and ETP=cluster when backend pods are also served by EgressIP` is running on SGW mode both v4 & v6

**- Description for the changelog**
`Support ETP=local with EIP`

/label ci

tssurya commented:
TODO: Add tests.

tssurya commented:
> TODO: Add tests.

done! let's see if the CI passes, besides fixing upgrade updates we are looking good here. Since we don't have indexes for LRPs and we rely on the match, I'd need to sort of sync cleanup all stale pod routes during upgrades to update to the new ones ugh.

tssurya commented:
@trozet PTAL this one is ready!

tssurya commented:
I tested the secondary networks EIP test case on lgw and it works as expected with this PR with no changes:

service traffic via eth1:
```
root@ovn-worker:/# conntrack -E | grep 36363                                                                                                          
    [NEW] tcp      6 120 SYN_SENT src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 [UNREPLIED] src=169.254.169.3 dst=172.20.0.2 sport=30432 dport=36363
    [NEW] tcp      6 120 SYN_SENT src=172.20.0.2 dst=169.254.169.3 sport=36363 dport=30432 [UNREPLIED] src=10.244.1.3 dst=172.20.0.2 sport=80 dport=36363 mark=2 zone=20
    [NEW] tcp      6 120 SYN_SENT src=172.20.0.2 dst=10.244.1.3 sport=36363 dport=80 [UNREPLIED] src=10.244.1.3 dst=172.20.0.2 sport=80 dport=36363 zone=24
 [UPDATE] tcp      6 60 SYN_RECV src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 src=169.254.169.3 dst=172.20.0.2 sport=30432 dport=36363    
 [UPDATE] tcp      6 432000 ESTABLISHED src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 src=169.254.169.3 dst=172.20.0.2 sport=30432 dport=36363 [ASSURED]
    [NEW] tcp      6 300 ESTABLISHED src=172.20.0.3 dst=172.20.0.2 sport=30432 dport=36363 [UNREPLIED] src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 mark=2 zone=6400
0
 [UPDATE] tcp      6 120 FIN_WAIT src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 src=169.254.169.3 dst=172.20.0.2 sport=30432 dport=36363 [ASSURED]
 [UPDATE] tcp      6 30 LAST_ACK src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 src=169.254.169.3 dst=172.20.0.2 sport=30432 dport=36363 [ASSURED]
 [UPDATE] tcp      6 120 TIME_WAIT src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 src=169.254.169.3 dst=172.20.0.2 sport=30432 dport=36363 [ASSURED]
[DESTROY] tcp      6 TIME_WAIT src=172.20.0.3 dst=172.20.0.2 sport=30432 dport=36363 src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 mark=2 zone=64000
[DESTROY] tcp      6 TIME_WAIT src=172.20.0.2 dst=172.20.0.3 sport=36363 dport=30432 src=169.254.169.3 dst=172.20.0.2 sport=30432 dport=36363 [ASSURED]
[DESTROY] tcp      6 TIME_WAIT src=172.20.0.2 dst=169.254.169.3 sport=36363 dport=30432 src=10.244.1.3 dst=172.20.0.2 sport=80 dport=36363 [ASSURED] mark=2 zone=20
[DESTROY] tcp      6 TIME_WAIT src=172.20.0.2 dst=10.244.1.3 sport=36363 dport=80 src=10.244.1.3 dst=172.20.0.2 sport=80 dport=36363 [ASSURED] zone=24
```

EIP traffic from pod to npclient:
```
root@ovn-worker:/# conntrack -E | grep icmp 
    [NEW] icmp     1 30 src=10.244.1.3 dst=172.20.0.2 type=8 code=0 id=7424 [UNREPLIED] src=172.20.0.2 dst=10.244.1.3 type=0 code=0 id=7424 zone=24
    [NEW] icmp     1 30 src=10.244.1.3 dst=172.20.0.2 type=8 code=0 id=7424 [UNREPLIED] src=172.20.0.2 dst=10.244.1.3 type=0 code=0 id=7424 zone=20
    [NEW] icmp     1 30 src=10.244.1.3 dst=172.20.0.2 type=8 code=0 id=7424 [UNREPLIED] src=172.20.0.2 dst=172.20.0.10 type=0 code=0 id=7424
    [NEW] icmp     1 30 src=172.20.0.10 dst=172.20.0.2 type=8 code=0 id=7424 [UNREPLIED] src=172.20.0.2 dst=172.20.0.10 type=0 code=0 id=7424 mark=2 zone=64000
 [UPDATE] icmp     1 30 src=10.244.1.3 dst=172.20.0.2 type=8 code=0 id=7424 src=172.20.0.2 dst=172.20.0.10 type=0 code=0 id=7424
```

I need to add an e2e test coverage for this.

tssurya commented on go-controller/pkg/ovn/egressip.go:
I am testing to see if upon upgrades old LRPs get removed or not.

tssurya replied:
adding some unit tests for that one..

trozet commented:
Like we discussed on slack, I think we can simplify by changing the qos rule to match on ct.rpl traffic rather than ! ct.rpl traffic, then add a single LRP that matches on pkt.mark==42 allow

tssurya commented:
yayyyy I finally fixed all those 72 unit tests and my e2es are looking good

tssurya commented:
/kind ci

tssurya commented:
/label ci

tssurya commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -818,8 +818,14 @@ func (h *defaultNetworkControllerEventHandler) AddResource(obj interface{}, from
 		h.oc.eIPC.nodeZoneState.UnlockKey(node.Name)
+		// add the 103 qos rule to new node's switch (we don't need to do it for node updates)
+		// (we don't need to remove this on node delete since entire node switch will get cleaned up anways)
+		err := h.oc.ensureDefaultNoRerouteQoSRules(node)
```
that's a great catch! and totally possible this happens (I kinda had this thought to be honest and then thought I was over engineering :) ; thanks for keeping me honest here). I'll add it to update as well although I won't have a way to check if alreadydone don't try, I guess on libovsdb side it will be a no-op most of the times.

tssurya replied:
I fixed this to also be called from update. I added a cachelookup. ptal if that works

trozet replied:
a slightly more efficient way would be to use a internal cache like we do for factory.NodeType, where we check if sync is needed by storing failures into a map:

```
if h.oc.isLocalZoneNode(oldNode) {
				// determine what actually changed in this update
				_, nodeSync := h.oc.addNodeFailed.Load(newNode.Name)
```

This cache lookup should be slightly faster than libovsdb cache look up, and avoids taking the nodeUpdateMutex. It might not be fast enough to matter...your call if you want to change it or not.

tssurya replied:
yeah agreed your suggestion is neater and that's how I guess we handle it elsewhere, for now just going with the mutex lock (opened FUP issue to get back to improving this: https://github.com/ovn-org/ovn-kubernetes/issues/4361)

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1336,13 +1350,26 @@ func (oc *DefaultNetworkController) initClusterEgressPolicies(nodes []interface{
 		return err
 	}
-
+	qoses, ops, err := createDefaultReRouteQoSRuleOps(oc.nbClient, oc.addressSetFactory, oc.controllerName)
```
hmm good point, I think its redundant like you said, removing this

tssurya replied:
done!

tssurya commented on go-controller/pkg/types/const.go:
```diff
@@ -94,6 +94,7 @@ const (
 	DefaultNoRereoutePriority             = 102
 	EgressSVCReroutePriority              = 101
 	EgressIPReroutePriority               = 100
+	EgressIPRerouteQoSRulePriority        = 103
```
> what made you pick 103 here? Does this conflict at all with EgressQoS also applied to the pod? 

no conflicts, 103 is a random number and the only reason I use a unique number is because I made the priority a "key" in the dbIndex for this QoS Rule. so its nice for me to get predicate based on priority and I know its an EIP owned one not EQoS owned one (although I also give it a unique name/family id keys)

tssurya commented:
@trozet I had to add client side db indexes for QoSRules because when two nodes were getting added at the same time in nonIC env, we were ending up inserting duplicate QoS Rules because nodes are handled in multiple threads and so two threads were trying to insert qos rules at the same time leading to multiple predicates found libovsdb error. 
```
2024/05/14 09:14:19 database/transaction/cache: "caller"={"file"="cache.go" "line"=1189} "level"=5 "msg"="inserting model" "table"="QoS" "uuid"="e3c08adb-fa2d-49d7-81e1-2891
5436db3e" "model"={"UUID"="e3c08adb-fa2d-49d7-81e1-28915436db3e" "Action"={"mark"=42} "Bandwidth"={} "Direction"="from-lport" "ExternalIDs"={"k8s.ovn.org/name"="EgressIP-Mar
k-Reply-Traffic" "priority"="103" "ip-family"="ip4" "k8s.ovn.org/id"="default-network-controller:EgressIP:103:EgressIP-Mark-Reply-Traffic:ip4" "k8s.ovn.org/owner-controller"
="default-network-controller" "k8s.ovn.org/owner-type"="EgressIP"} "Match"="ip4.src == $a4548040316634674295 && ct.trk && ct.rpl" "Priority"=103} 

2024/05/14 09:14:19 database/transaction/cache: "caller"={"file"="cache.go" "line"=1196} "level"=5 "msg"="updating model" "table"="Logical_Switch" "uuid"="74916f3a-7f7a-4701
-b1c5-90212493ad1b" "old"={"UUID"="74916f3a-7f7a-4701-b1c5-90212493ad1b" "ACLs"=[] "Copp"=null "DNSRecords"=[] "ExternalIDs"={} "ForwardingGroups"=[] "LoadBalancer"=[] "Load
BalancerGroup"=[] "Name"="node2" "OtherConfig"={} "Ports"=[] "QOSRules"=[]} "new"={"UUID"="74916f3a-7f7a-4701-b1c5-90212493ad1b" "ACLs"=[] "Copp"=null "DNSRecords"=[] "Exter
nalIDs"={} "ForwardingGroups"=[] "LoadBalancer"=[] "LoadBalancerGroup"=[] "Name"="node2" "OtherConfig"={} "Ports"=[] "QOSRules"=["e3c08adb-fa2d-49d7-81e1-28915436db3e"]}

2024/05/14 09:14:19 cache: "caller"={"file"="cache.go" "line"=1189} "level"=5 "msg"="inserting model" "table"="QoS" "uuid"="e3c08adb-fa2d-49d7-81e1-28915436db3e" "model"={"U
UID"="e3c08adb-fa2d-49d7-81e1-28915436db3e" "Action"={"mark"=42} "Bandwidth"={} "Direction"="from-lport" "ExternalIDs"={"ip-family"="ip4" "k8s.ovn.org/id"="default-network-c
ontroller:EgressIP:103:EgressIP-Mark-Reply-Traffic:ip4" "k8s.ovn.org/owner-controller"="default-network-controller" "k8s.ovn.org/owner-type"="EgressIP" "k8s.ovn.org/name"="E
gressIP-Mark-Reply-Traffic" "priority"="103"} "Match"="ip4.src == $a4548040316634674295 && ct.trk && ct.rpl" "Priority"=103}

2024/05/14 09:14:19 cache: "caller"={"file"="cache.go" "line"=1196} "level"=5 "msg"="updating model" "table"="Logical_Switch" "uuid"="74916f3a-7f7a-4701-b1c5-90212493ad1b" "
old"={"UUID"="74916f3a-7f7a-4701-b1c5-90212493ad1b" "ACLs"=[] "Copp"=null "DNSRecords"=[] "ExternalIDs"={} "ForwardingGroups"=[] "LoadBalancer"=[] "LoadBalancerGroup"=[] "Na
me"="node2" "OtherConfig"={} "Ports"=[] "QOSRules"=[]} "new"={"UUID"="74916f3a-7f7a-4701-b1c5-90212493ad1b" "ACLs"=[] "Copp"=null "DNSRecords"=[] "ExternalIDs"={} "Forwardin
gGroups"=[] "LoadBalancer"=[] "LoadBalancerGroup"=[] "Name"="node2" "OtherConfig"={} "Ports"=[] "QOSRules"=["e3c08adb-fa2d-49d7-81e1-28915436db3e"]}

2024/05/14 09:14:19 database/transaction/cache: "caller"={"file"="cache.go" "line"=1189} "level"=5 "msg"="inserting model" "table"="QoS" "uuid"="9d5d02e7-d269-4d4f-bfb4-6610
d668d935" "model"={"UUID"="9d5d02e7-d269-4d4f-bfb4-6610d668d935" "Action"={"mark"=42} "Bandwidth"={} "Direction"="from-lport" "ExternalIDs"={"ip-family"="ip4" "k8s.ovn.org/n
ame"="EgressIP-Mark-Reply-Traffic" "k8s.ovn.org/id"="default-network-controller:EgressIP:103:EgressIP-Mark-Reply-Traffic:ip4" "k8s.ovn.org/owner-controller"="default-network
-controller" "k8s.ovn.org/owner-type"="EgressIP" "priority"="103"} "Match"="ip4.src == $a4548040316634674295 && ct.trk && ct.rpl" "Priority"=103}

2024/05/14 09:14:19 database/transaction/cache: "caller"={"file"="cache.go" "line"=1196} "level"=5 "msg"="updating model" "table"="Logical_Switch" "uuid"="a886b025-3d2f-4b6e
-a198-5b04a1ba5301" "old"={"UUID"="a886b025-3d2f-4b6e-a198-5b04a1ba5301" "ACLs"=[] "Copp"=null "DNSRecords"=[] "ExternalIDs"={} "ForwardingGroups"=[] "LoadBalancer"=[] "Load
BalancerGroup"=[] "Name"="node1" "OtherConfig"={} "Ports"=[] "QOSRules"=[]} "new"={"UUID"="a886b025-3d2f-4b6e-a198-5b04a1ba5301" "ACLs"=[] "Copp"=null "DNSRecords"=[] "Exter
nalIDs"={} "ForwardingGroups"=[] "LoadBalancer"=[] "LoadBalancerGroup"=[] "Name"="node1" "OtherConfig"={} "Ports"=[] "QOSRules"=["9d5d02e7-d269-4d4f-bfb4-6610d668d935"]}

2024/05/14 09:14:19 cache: "caller"={"file"="cache.go" "line"=1189} "level"=5 "msg"="inserting model" "table"="QoS" "uuid"="9d5d02e7-d269-4d4f-bfb4-6610d668d935" "model"={"U
UID"="9d5d02e7-d269-4d4f-bfb4-6610d668d935" "Action"={"mark"=42} "Bandwidth"={} "Direction"="from-lport" "ExternalIDs"={"k8s.ovn.org/name"="EgressIP-Mark-Reply-Traffic" "k8s
.ovn.org/id"="default-network-controller:EgressIP:103:EgressIP-Mark-Reply-Traffic:ip4" "k8s.ovn.org/owner-controller"="default-network-controller" "k8s.ovn.org/owner-type"="
EgressIP" "priority"="103" "ip-family"="ip4"} "Match"="ip4.src == $a4548040316634674295 && ct.trk && ct.rpl" "Priority"=103}

2024/05/14 09:14:19 cache: "caller"={"file"="cache.go" "line"=1196} "level"=5 "msg"="updating model" "table"="Logical_Switch" "uuid"="a886b025-3d2f-4b6e-a198-5b04a1ba5301" "
old"={"UUID"="a886b025-3d2f-4b6e-a198-5b04a1ba5301" "ACLs"=[] "Copp"=null "DNSRecords"=[] "ExternalIDs"={} "ForwardingGroups"=[] "LoadBalancer"=[] "LoadBalancerGroup"=[] "Na
me"="node1" "OtherConfig"={} "Ports"=[] "QOSRules"=[]} "new"={"UUID"="a886b025-3d2f-4b6e-a198-5b04a1ba5301" "ACLs"=[] "Copp"=null "DNSRecords"=[] "ExternalIDs"={} "Forwardin
gGroups"=[] "LoadBalancer"=[] "LoadBalancerGroup"=[] "Name"="node1" "OtherConfig"={} "Ports"=[] "QOSRules"=["9d5d02e7-d269-4d4f-bfb4-6610d668d935"]}
```
The only other feature using QoS rules on switch is EgressQoS in addition to EgressIPs. So introducing client side indexes means from this version onwards eqos feature will start creating qos rules in OVN with the new externalIDs that are unique per QoS Rule. That means when users upgrade from older versions to new versions with EgressQoS, there will be stale left overs which I don't plan to clean up in this PR. Anyways since EgressQoS feature is still not GA-ed and is only Dev Preview to the point where even the API may change in a breaking way including the name, I think its ok to start using db indexes for EQoS without worrying about updating older QoSRules i.e upgrades for QoS is not handled at the moment so stale QoS Rules will be left behind without the new externalIDs but that should be fine

tssurya commented:
adding indexes didn't work actually. The reason here is because the race was in the cache level, so since 2 threads try to insert at same time, we hit some hiccupps.. I have instead used lock now (I added a new commit for dbindexes, but we can also move that out of this PR if that's too much)

tssurya commented:
@trozet : I have addressed your comments. PTAL
As I was addressing comment1: https://github.com/ovn-org/ovn-kubernetes/pull/4265#discussion_r1587706794 where I was creating the QoS rules globally on startup and now I am only doing it from node add events; I ended up with a race condition where for two node adds we were trying to create the qos rules at the same time from two different threads and due to libovsdb cache lag ended up with multiple objects, so I am not using the `nodeUpdateMutex` to lock and perform the updates. I have also added the logic for cache lookup like we discussed yesterday. Those are the new changes in the recent push into commit: https://github.com/ovn-org/ovn-kubernetes/pull/4265/commits/70c800c2fc7bc2e658fbe40451575f18087411a1

Unintentionally I added the dbindexes commit for eqos: https://github.com/ovn-org/ovn-kubernetes/pull/4265/commits/07fa4d1f493bfca544a407e610273ad7a4031350 which is probably not really needed ATM; if you prefer to keep that commit as separate PR I can cut it out.

trozet commented on go-controller/pkg/libovsdb/libovsdb.go:
```diff
@@ -203,6 +203,7 @@ func NewNBClientWithConfig(cfg config.OvnAuthConfig, promRegistry prometheus.Reg
 		nbdb.LoadBalancerTable:  {{Columns: []model.ColumnKey{{Column: "name"}}}},
 		nbdb.LogicalSwitchTable: {{Columns: []model.ColumnKey{{Column: "name"}}}},
 		nbdb.LogicalRouterTable: {{Columns: []model.ColumnKey{{Column: "name"}}}},
+		nbdb.QoSTable:           {{Columns: []model.ColumnKey{{Column: "external_ids", Key: types.PrimaryIDKey}}}},
```
looks like address set type is also missing...why?

tssurya replied:
so I cannot do LogicalRouterPolicy yet because the rest of the LRPs do not yet have externalIDs so that will need to be a refactor in a separate PR change. However when we do add new policies we should start using externalIDs so that the amount of conversion we have to do is less

re: address-set... hmm no idea I'm pretty sure Nadia converted all of them to use externalIDs.. @npinaeva do you know why AddressSets are not in this list for client side indexing using primaryIDs?

trozet replied:
ok so the rule of thumb here is we can only add client indexes once all objects have been converted to db ids...makes sense. Let's wait for @npinaeva to give us the status on address sets.

tssurya replied:
yeah my guess is going to be that so far all address-sets have unique names and name is the indexing we use today which probably suffices, but let's wait to see what Nadia says because the whole point of changing to use indexes was for this..

npinaeva replied:
that is because address set name is a server-side index and it is based on primary id https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/address_set/address_set.go#L263-L265
all address set operations are done via address set factory that makes sure the name is always set.
For port groups it is the same https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/libovsdb/util/port_group.go#L10
anything that doesn't have a primaryID-based server-side index should leverage client side indexing like you did here.

tssurya replied:
oh! that's good to know; I think we should have docs for indexing OR even a broader set of docs around libovsdb in our contributor docs; I'll open an issue to track it

1. client side indexing
2. server side indexing
3. how ovnkube uses libovsdb
4. moving forward all new OVN Constructs must leverage dbindexes
5. details of indexing nuances of each ovn object in ovnkube etc


We'll see if we can find cycles during our documentation-workshop-friday if we can plan such a thing :D

opened https://github.com/ovn-org/ovn-kubernetes/issues/4362 and resolving this comment.

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1996,9 +2014,118 @@ func createDefaultNoReroutePodPolicies(nbClient libovsdbclient.Client, v4Cluster
+		qosExists = true
+	}
+	if !qosExists {
+		existingQoSes, ops, err = createDefaultReRouteQoSRuleOps(oc.nbClient, oc.addressSetFactory, oc.controllerName)
```
why do we create the ops if the node might not be local zone? could we skip calling this entire function if the node is remote?

tssurya replied:
oh good point, my check seems too late in cycle I should probably push this to when we get the event in default network controller

tssurya replied:
done.
